src/capture/capture_video.cpp

```diff
@@ -358,6 +358,42 @@ static void create_avi_file(const uint16_t width, const uint16_t height,
 	video.audio.bytes_written   = 0;
 }
 
+// Reads the image's pixels from the emulated DOS video RAM (which is always
+// little-endian) and passes them in the same byte-order to the encoder.
+//
+// Ideally this could just be one memcpy, but many DOS video modes need their
+// pixels either doubled horizontally and/or vertically to produce the correct
+// on-screen size. So extra copies are needed in both these cases.
+//
+static void compress_frame(const RenderedImage& image)
+{
+	const auto& src = image.params;
+
+	auto src_row = image.image_data;
+	uint8_t tgt_row[SCALER_MAXWIDTH * 4] = {};
+
+	const auto pixel_multiple = (src.double_width ? 2 : 1);
+	const auto row_multiple   = (src.double_height ? 2 : 1);
+
+	const auto src_bpp = to_bytes_per_pixel(src.pixel_format);
+	const auto tgt_bpp = to_bytes_per_pixel(to_zmbv_format(src.pixel_format));
+
+	// Arrange the source bytes into a target row before compressing
+	for (auto i = 0; i < src.height; ++i, src_row += image.pitch) {
+		auto src_pixel = src_row;
+		auto tgt_pixel = tgt_row;
+		for (auto j = 0; j < src.width; ++j, src_pixel += src_bpp) {
+			for (auto k = 0; k < pixel_multiple; ++k, tgt_pixel += tgt_bpp) {
+				std::memcpy(tgt_pixel, src_pixel, src_bpp);
+			}
+		}
+		for (auto j = 0; j < row_multiple; ++j) {
+			const uint8_t* row_pointer = tgt_row;
+			video.codec->CompressLines(1, &row_pointer);
+		}
+	}
+}
+
```

Comment:
The function is identical in loop structure and quantities (num rows, num columns, optional doubling on width, optional doubling on height) as the prior one.

It produces the exact same results on little-endian as `main`.

The only difference is how it copies the values:
 - The prior code used pointer casts (buint16*, uint32*)
 - The new one copies byte-by-byte.

It's impossible to get different results in this branch on little endian.
 - All AVIs will be identical
 - All pngs (Alt+F5 and Ctrl+F5) will be identical

The loop is just a simplified and more generic form of the prior.

I'll compare them side-by-side and draw arrows to the elements if it helps :-)

Reply:
I'll add a comment in the commit to make it clearer that the loop structure and logic is identical to the existing function, with the only change being the way the pixels are copied from the source (using memcopy byte-by-byte instead of multi-byte pointers).

Reply:
Yeah a comment in the git commit would be good. I admittedly did not spend the time to correlate the removed code with the added code 😄